Blockade mechanics

vegeta1k95 opened the pull request:
This PR includes implementation for tile/city blockade mechanics from Civ 5: 

Blocked tile = tile which (this turn) does not produce yields, cannot be worked/locked.

**Mechanics:**
1) Enemy military land unit blocks tile it is on  (includes embarked);
2) Enemy military naval unit blocks tile it is on + adjacent water tiles which do not have friendly/neutral military units on them;
3) If coastal city has **all** adjacent water tiles blocked - city loses it's _harbor_ city connection to the capital;

This includes the following change: if tile was worked/locked and became blocked, city removes lock and reassigns population accordingly.

Looks as follows:
1) Enemy infantry block tiles they are standing on:
![image](https://user-images.githubusercontent.com/32207817/218227928-40ae0a09-1094-4235-a22d-d142e72b51f9.png)

2) Enemy Destroyer blocks every adjacent tile, except the one where my Carrier is located:
![image](https://user-images.githubusercontent.com/32207817/218232243-84bc2169-a5a6-4e8b-b378-e80e36e7adcb.png)

AdityaMH commented:
It's time to see hidden submarine freely.

vegeta1k95 commented:
> It's time to see hidden submarine freely.

That how it was in Civ 5, though it is debatable whether we should implement it the same.
Adding visibility check for units and a matter of one line of code, but it should be decided beforehand.

vegeta1k95 commented:
Added City blockade mechanics

yairm210 commented on core/src/com/unciv/logic/map/tile/Tile.kt:
```diff
@@ -429,6 +429,40 @@ open class Tile : IsPartOfGameInfoSerialization {
+
+        // If tile has unit
+        if (unit != null) {
+            return when {
```
Can if/else

vegeta1k95 replied:
Don't understand which `if/else` you are talking about here.

yairm210 replied:
My mistake, I thought the when had only 2 possible outcomes, not 3

yairm210 commented on core/src/com/unciv/ui/utils/UnitGroup.kt:
```diff
@@ -125,7 +126,9 @@ class UnitGroup(val unit: MapUnit, val size: Float): Group() {
 
         if (unit.health < 100) { // add health bar
-            addActor(ImageGetter.getHealthBar(unit.health.toFloat(), 100f, size))
+            val hp = ImageGetter.getHealthBar(unit.health.toFloat(), 100f, size*0.78f)
```
What's the magic 0.78?

vegeta1k95 replied:
This is similar multiplier to FlagBackground `innerMultiplier`. Only 0.1 smaller, so that HP bar width is a bit smaller than flag's inner circle

yairm210 commented:
Looks good, but since this is an entirely be mechanic it needs a tutorial, which should pop up the first time one of your tiles is blockaded

vegeta1k95 commented:
@yairm210 
Added tutorials: 
1) For tile blockade, on CityScreen
2) For city blockade, on WorldScreen

Added city blockade status icon for city plates:
![image](https://user-images.githubusercontent.com/32207817/218614628-1494e06b-466c-4e96-9f0c-597eed9bcadc.png)

Caballero-Arepa commented:
Amazing!
Cant wait to declare war on my firends xD

> For instance, Civ 5 Vox Populi/Community Patch allows blockade only by visible units (so subs do not blockade).

I dont think so, Germany blockaded Britain in both World Wars with Submarines, and Rusia is doing it in Ukraine, so it is realistic.
